Send frame notifications even if there is no buffer

wmww opened the pull request:
* The frame notifications are copied by value into the lamda now, and there is no need to hold them in the surface. (Does this screw stuff up? Was there a reason we were maintaining a list in the surface before, or was that legacy?)
* Frame notifications are now sent even if there is no buffer or if the buffer is null (this is consistent with Weston)

This fixes the gedit lockup when subsurfaces are dismissed.

AlanGriffiths commented:
> The frame notifications are copied by value into the lamda now, and there is no need to hold them in the surface. (Does this screw stuff up? Was there a reason we were maintaining a list in the surface before, or was that legacy?)

Empirically, some clients needed  notification for all frames submitted at the time of composition, not just those submitted at the time the buffer was committed. (IIRC you tried changing this once before and things broke.)

I'll do some testing.

wmww commented:
> IIRC you tried changing this once before and things broke.

It's all coming back to me.

The diff was pretty short, and mostly needed to be changed, so I rebased out the bad commit instead of adding a revert commit. I hope it's now a bit cleaner with a proper function instead of multiple lambdas. Also, I added a comment so hopefully I'll stop breaking this now. I've tested it with gedit, kate and that XWayland libreoffice command (which I confirmed to be broken previously). All seem to work fine.
